Sage/carbon
# fix(menu): make submenu scrollable when inside a fixed navigation bar

### Proposed behaviour

When a submenu is used inside a NavigationBar with position="fixed", or inside a GlobalHeader (which always has position fixed), and the available screen space is too small to hold the entire menu, the submenu should become scrollable so that, whether using the mouse or keyboard, users can easily navigate to all the submenu items.



### Current behaviour

When a submenu is used inside a navigation bar with position fixed, and it is too long to fit in the remaining screen space, it is not practically possible to reach the menu items at the bottom. When using a mouse, there is no scrollbar on the submenu and scrolling down the page as a whole doesn't help as the whole thing is fixed relative to the screen. When using the keyboard, focus technically goes to the off-screen menu items but as a user you cannot see what you would be pressing/activating.



### Checklist




#### QA


### Additional context


Note that if using a ScrollableBlock inside a submenu where this new feature is used, the result will be a double-scrollbar. This has been previously approved by @harpalsingh as acceptable, so I have not specifically tested this scenario and have not put the PR down for UX QA.

### Testing instructions



Check that the new funcitonality works as intended inside both GlobalHeader and NavigationBar, with all combinations of props (a test story has been added for each of these, and that for NavigationBar has args which can be adjusted locally to try out different scenarios).

Also do full regression testing of both components as part of this fix involved a small refactor of both of them.

Ensure all situations with submenu - including with ScrollableBlock - behave as expected and have not regressed.

Codesandbox demo of the original issue inside a GlobalHeader (try resizing the screen and reopening the menu, particularly with the fixed version, to ensure it works no matter what the screen size): https://codesandbox.io/s/inspiring-smoke-3hsmmg?file=/src/App.tsx
